File: config/deploy.rb
require 'mina/bundler'
require 'mina/rails'
require 'mina/git'

account = 'rochedor'

set :repository, 'git@gitlab.com:logomotion/rochedor-nsi'
set :branch, 'dev' if ENV['on'] == 'dev'
set :branch, 'staging' if ENV['on'] == 'staging'

set :domain, 'rochedor.fr'
set :user, 'root'

set :forward_agent, true
set :term_mode, nil

set :deploy_to, "/home/#{account}/nsi_staging" if ENV['on'] == 'staging'
set :deploy_to, "/home/#{account}/nsi_dev" if ENV['on'] == 'dev'

set :shared_dirs, ['vendor', 'web/uploads', 'web/biblio', 'web/kiwi', 'web/site', 'var/cache', 'var/logs']
set :shared_files, ['web/.htaccess', 'web/robots.txt', 'web/kiwi.php', 'web/adminer-lrdo.php', 'app/config/parameters.yml']

task :deploy => :environment do
	deploy do
		invoke :'git:clone'
		invoke :'deploy:cleanup'
		invoke :'deploy:link_shared_paths'
		on :launch do
			command "export SYMFONY_ENV=prod"
			command "ea-php72 composer.phar install --optimize-autoloader"
			command "ea-php72 composer.phar dump-autoload --optimize --classmap-authoritative"
			command "/scripts/restartsrv_apache_php_fpm"
			command "yarn install"
			command "ea-php72 bin/console assets:install --env=prod"
			command "ea-php72 bin/console cache:clear --env=prod"
			command "chmod +x ./node_modules/.bin/encore"
			command "./node_modules/.bin/encore production --config-name configDev" if ENV['on'] == 'dev' || ENV['on'] == 'staging'
			command "./node_modules/.bin/encore production --config-name configProd" if ENV['on'] == 'prod'
			command "find #{fetch(:deploy_to)} -type d -exec chmod 755 {} +"
			command "find #{fetch(:deploy_to)} -type f -exec chmod 644 {} +"
			command "chown -R #{account}:#{account} #{fetch(:deploy_to)}"
			command "ea-php72 bin/console doctrine:schema:update --dump-sql > ../../update.sql"
		end
	end
end

File: deploy/deploy.php
<?php

    namespace Deployer;
    
    require 'recipe/composer.php';
    require 'recipe/common.php';

    set('repository', 'git@gitlab.com:logomotion/rochedor-nsi.git');
    set('git_tty', false);
    set('default_stage', 'dev');
    set('shared_files', ['web/.htaccess', 'web/robots.txt', 'app/config/parameters.yml']);
    set('shared_dirs', ['vendor', 'web/uploads', 'var/cache', 'var/logs', 'node_modules', 'web/biblio']);
    set('bin/php', '/usr/local/bin/php -c deploy/deploy.ini');

    host('staging.rochedor.fr')
        ->stage('staging')
        ->user('root')
        ->hostname('54.36.109.76')
        ->set('account_dir', 'rochedor')
        ->set('branch', 'staging')
        ->set('encore_config_name', 'configDev')
        ->set('deploy_path', '/home/{{account_dir}}/public_html/nsi/src_staging');

    host('rochedor.fr')
        ->stage('prod')
        ->user('root')
        ->hostname('54.36.109.76')
        ->set('account_dir', 'rochedor')
        ->set('branch', 'prod')
        ->set('encore_config_name', 'configDev')
        ->set('deploy_path', '/home/{{account_dir}}/public_html/nsi/src_prod');

    task('deploy', [
        'deploy:info',
        'deploy:prepare',
        'deploy:lock',
        'deploy:release',
        'deploy:update_code',
        'deploy:shared',
        'deploy:writable',
        'deploy:clear_paths',
        'deploy:symlink',
        'deploy:unlock',
        'install',
        'permissions',
        'project-cleanup',
        'cleanup',
        'success'
    ]);

    task('project-cleanup', function () {
        run('rm -rf script/');
        run('rm -rf deploy/');
        run('rm -f docker-compose.yml Dockerfile settings.json .gitlab-ci.yml .gitignore .env-dist');
    });

    task('install', function () {
        run('export SYMFONY_ENV=prod');
        run('cd {{release_path}} && {{bin/php}} composer.phar install --optimize-autoloader');
        run('cd {{release_path}} && {{bin/php}} composer.phar dump-autoload --optimize --classmap-authoritative');
        run('cd {{release_path}} && yarn install');
        run('cd {{release_path}} && {{bin/php}} bin/console cache:clear --env=prod');
        run('cd {{release_path}} && {{bin/php}} bin/console assets:install --env=prod');
        run('cd {{release_path}} && chmod +x ./node_modules/.bin/encore');
        run('cd {{release_path}} && ./node_modules/.bin/encore production --config-name {{encore_config_name}}');
        run('cd {{release_path}} && {{bin/php}} bin/console doctrine:schema:update --dump-sql > ../../update.sql');
    });

    task('permissions', function () {
        run('find {{deploy_path}} -type d -exec chmod 755 {} +');
        run('find {{deploy_path}} -type f -exec chmod 644 {} +');
        run('chown -R {{account_dir}}:{{account_dir}} {{deploy_path}}');
    });

    after('deploy:failed', 'deploy:unlock');
